### PR TITLE
Api fetching insights

### PR DESCRIPTION
**Pull request by tspenov**
#### Summary

card: https://favro.com/card/bdbb4f24afc48b29c74f4fa4/San-1467

- [x] All insights API query for anon and logged in users
- [x] GQL query API for all insights by user
- [x] GQL query API for all insights a user has voted for
- [x] authorize fetching single Post
- [x] move post functionality fo creating, deleting posts in post_resolver 
- [x] use dataloader to prevent N+1 queries when `totalSanVotes` or `votedAt` are requested
Dataloader is not used directly but votes are preloaded do we shouldn't have N+1 queries for `votedAt` and `totalSanVotes` fields

1. All insights
1.1 when not authenticated only `title`, `shortDesc`. 
```
{
  allInsights {
    title,
    shortDesc
  }
}
```

Response:
```
{
  "data": {
    "allInsights": [
      {
        "title": "Checksd",
        "shortDesc": null
      },
      {
        "title": "Example title",
        "shortDesc": null
      },
     ...
    ]
  }
}
```

1.1 when authenticated
request:
```
{
  allInsights {
    id,
    createdAt,
    images {
    	imageUrl
    },
    link,
    moderationComment,
    title,
    text,
    state,
    shortDesc,
    user {
      email
    },
    relatedProjects {
      ticker
    }
  }
}
```
response:
```
{
  "data": {
    "allInsights": [
      {
        "user": {
          "email": "yura@santiment.net"
        },
        "title": "fghdgh",
        "text": "gfhfhgf\njgfhjghkj ![3b2ar](https://stage-sanbase-images.s3.amazonaws.com/uploads/2dffecb0ea60e6216a2e2b1467164f2a99e6eccc2761a16aeab9f675965a36c8_1522206432787_unknown.png)\n\ngfhjgj\n",
        "state": null,
        "shortDesc": null,
        "relatedProjects": [],
        "moderationComment": null,
        "link": null,
        "images": [],
        "id": "97",
        "createdAt": "2018-03-28T03:07:23Z"
      },
    ...
    ]
  }
}
```

2. All insights for user

request:
```
{
  allInsightsForUser(userId: 2) {
    id,
    createdAt,
    images {
    	imageUrl
    },
    link,
    moderationComment,
    title,
    text,
    state,
    shortDesc,
    user {
      id,
      email
    },
    relatedProjects {
      ticker
    }
  }
}
```

response:

```
{
  "data": {
    "allInsightsForUser": [
      {
        "user": {
          "id": "2",
          "email": "yura@santiment.net"
        },
        "title": "Checksd",
        "text": null,
        "state": null,
        "shortDesc": null,
        "relatedProjects": [],
        "moderationComment": null,
        "link": "https://asdf.ri",
        "images": [],
        "id": "37",
        "createdAt": "2018-02-16T06:46:37Z"
      },
     ...
    ]
  }
}
```

3. All insights a given user has voted for
request:

```
{
  allInsightsUserVoted(userId: 2) {
    id,
    createdAt,
    images {
    	imageUrl
    },
    link,
    moderationComment,
    title,
    text,
    state,
    shortDesc,
    user {
      id,
      email
    },
    relatedProjects {
      ticker
    }
  }
}
```

response:
```
{
  "data": {
    "allInsightsUserVoted": [
      {
        "user": {
          "id": "17",
          "email": "yura.z@santiment.net"
        },
        "title": "asdjfhakj adsfasd f",
        "text": "asdfjhlkdshf\nasdfkjahldf\n\n\nasdkfjhk\nsdkfjh ![847vf](https://stage-sanbase-images.s3.amazonaws.com/uploads/00663344fed8f3e62e3711978d63a42e14f5c5a2e9252067da6f381ccb0255c0_1521260466305_sanbase-chart-OMG-2018-02-23T01_46_29+05_00.jpg)\n\nadskjflahdf\n\nasdkfjhaldkfj",
        "state": null,
        "shortDesc": null,
        "relatedProjects": [],
        "moderationComment": null,
        "link": null,
        "images": [],
        "id": "93",
        "createdAt": "2018-03-17T04:21:28Z"
      },
    ....
    ]
  }
}
```

4. Fetch single insight
request:
```
{
  post(id: 92) {
    id,
    createdAt,
    images {
    	imageUrl
    },
    link,
    moderationComment,
    title,
    text,
    state,
    shortDesc,
    user {
      id,
      email
    },
    relatedProjects {
      ticker
    }
  }
}
```
response:
```
{
  "data": {
    "post": {
      "user": {
        "id": "17",
        "email": "yura.z@santiment.net"
      },
      "title": "Check 3",
      "text": "asdfjh\n\nasdf\n\n\nasdf\n\n ![1f9qn](https://stage-sanbase-images.s3.amazonaws.com/uploads/2a3e17e11db9d97ec49a4c0a1b7a00949406333f0e300d72dc9807b5726b438e_1521258614546_29203205911_069adc67bb_k.jpg)\n\nasdf\n\nasdf\n\n\nasdf\n\ncheck\n\n",
      "state": null,
      "shortDesc": null,
      "relatedProjects": [],
      "moderationComment": null,
      "link": null,
      "images": [],
      "id": "92",
      "createdAt": "2018-03-17T03:50:31Z"
    }
  }
}
```

### CODE REVIEW

File: lib/sanbase_web/graphql/resolvers/post_resolver.ex [hunk 1]
```diff
@@ -0,0 +1,56 @@
+defmodule SanbaseWeb.Graphql.Resolvers.PostResolver do
+  require Logger
+
+  import Ecto.Query
+
+  alias Sanbase.Voting.Post
+  alias Sanbase.Repo
+
+  @preloaded_assoc [:votes, :user, :related_projects, :images]
```
Comment (tspenov) [inline lib/sanbase_web/graphql/resolvers/post_resolver.ex]: Any ideas what should we preloaded with the posts?

File: lib/sanbase_web/graphql/resolvers/post_resolver.ex [hunk 2]
```diff
@@ -0,0 +1,56 @@
+defmodule SanbaseWeb.Graphql.Resolvers.PostResolver do
```
Comment (IvanIvanoff) [inline lib/sanbase_web/graphql/resolvers/post_resolver.ex]: We now have the functionality for the posts in `voting_resolver` (which is probably not the best name)

**Comment by Partysun** [comment]
@tspenov Hi, Could you share examples of a request and an answer?

**Comment by Partysun** [comment]
@tspenov Upload new API on stage. I need this to change my side.

**Comment by tspenov** [comment]
@Partysun Added example requests and responses for the API. Will merge when approved

**Comment by tspenov** [comment]
@Partysun @valo Do we need some kind of pagination for the `allInsights` ?

File: lib/sanbase_web/graphql/resolvers/post_resolver.ex (hunk 2 above)
Comment (tspenov) [reply lib/sanbase_web/graphql/resolvers/post_resolver.ex]: @IvanIvanoff moved create/delete post to the post_resolver

**Comment by valo** [comment]
@tspenov I think we can roll without pagination for now. When we start getting a lot of posts we can figure out what kind of pagination will work (pages or infinite scroll)

**Pull request merged**